rmw_zenoh_cpp/src/rmw_init.cpp

```diff
@@ -242,148 +165,94 @@ rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
     idstr[sizeof(zid.id) * 2] = '\0';
     // TODO(yadunund): Can we get the size of the shm from the config even though it's not
     // a standard parameter?
-    context->impl->shm_manager =
+    shm_manager =
       zc_shm_manager_new(
-      z_loan(context->impl->session),
+      z_loan(session),
       idstr,
       SHM_BUFFER_SIZE_MB * 1024 * 1024);
-    if (!context->impl->shm_manager.has_value() ||
-      !zc_shm_manager_check(&context->impl->shm_manager.value()))
+    if (!shm_manager.has_value() ||
+      !zc_shm_manager_check(&shm_manager.value()))
     {
       RMW_SET_ERROR_MSG("Unable to create shm manager.");
       return RMW_RET_ERROR;
     }
   }
   auto free_shm_manager = rcpputils::make_scope_exit(
-    [context]() {
-      if (context->impl->shm_manager.has_value()) {
-        z_drop(z_move(context->impl->shm_manager.value()));
+    [&shm_manager]() {
+      if (shm_manager.has_value()) {
+        z_drop(z_move(shm_manager.value()));
       }
     });
 
   // Initialize the guard condition.
-  context->impl->graph_guard_condition =
+  rmw_guard_condition_t * graph_guard_condition =
```

Comment:
In the new structure, it seems like the `rmw_context_impl_s` structure is the "owner" of the `graph_guard_condition`.  If I'm correct about that, then my suggestion here is to move the "construction" of the graph guard_condition (this code, and the next dozen lines or so) into the constructor of `rmw_context_impl_s`.  I think it makes more sense there.  What do you think?

Reply:
Yeah I wanted to do that  but was unsure how to handle the case when we need to return `RMW_RET_BAD_ALLOC` when memory allocation fails. Should the constructor throw an error? In #259 I was able to rely on the `make()` pattern but I don't see a way to do that here unless we switch to a singleton of `rmw_context_impl_s`.

Reply:
I think the constructor can throw; we are careful to use `RMW_TRY_PLACEMENT_NEW` when constructing, which captures the error and turns it into a `return` through the C API.

Reply:
Ah that's pretty neat! Will update accordingly!

Reply:
After learning that `RMW_TRY_PLACEMENT_NEW` elegantly handles exceptions, I've updated the constructor here to also instantiate the Zenoh router and shm manager. https://github.com/ros2/rmw_zenoh/pull/258/commits/a6fb52136f499dd411de284a912ef0cd99b2e85f